Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -578,10 +579,23 @@ static bool needsEAGLOnMac()
     return allocateAndBindDisplayBufferBacking();
 }
 
+bool GraphicsContextGLCocoa::setDrawingBufferColorSpaceIfSupported(DestinationColorSpace colorSpace)
+{
+    if (m_drawingBufferColorSpace != colorSpace) {
+        m_drawingBufferColorSpace = colorSpace;
+
+        if (!getInternalFramebufferSize().isEmpty() && !reshapeDisplayBufferBacking()) {
+            RELEASE_LOG(WebGL, "Fatal: Unable to allocate backing store of size %d x %d", getInternalFramebufferSize().width(), getInternalFramebufferSize().height());
+            forceContextLost();
+        }
+    }
+    return true;
+}
+
 bool GraphicsContextGLCocoa::allocateAndBindDisplayBufferBacking()
 {
     ASSERT(!getInternalFramebufferSize().isEmpty());
-    auto backing = IOSurface::create(nullptr, getInternalFramebufferSize(), DestinationColorSpace::SRGB());
+    auto backing = IOSurface::create(nullptr, getInternalFramebufferSize(), m_drawingBufferColorSpace);
```

Comment:
when the buffer is sent for compositing previous buffer is, in some cases, recycled.
You probably need to address this in some way in `prepareForDisplay()`
E.g. 
Buffer 1: srgb
Buffer 2: p3
Buffer 3: reuse buffer 1, need to somehow address that it should be p3.

Reply:
I think reusable buffers are stored with their colorspace, so there shouldn't be discrepancies; but I'll double check...

Reply:
What I mean is that consider sequence of:
1. draw 1 frame with srgb
2. switch to p3
3. draw 1 frame with p3
3.1. Here we don't want the case where we pull a buffer recycled buffer from the swap chain and it ends up containing srgb

>I think reusable buffers are stored with their colorspace, so there shouldn't be discrepancies

That's the discrepancy: if we reuse a buffer that was submitted as srgb..

Reply:
I meant, I think the buffers are keyed on their colorspace, so when the caller requests a P3 buffer, it should get one that was created with a P3 colorspace. But I should double/triple-check!

Reply:
To expand:
Currently prepareForDisplay() checks that the resolution of the recycled buffer matches the current resolution of the context.

In future,  prepareForDisplay() should check that the resolution of the recycled buffer matches the current resolution of the context AND that the color space matches the current drawing buffer color space of the context.